### PR TITLE
Dockerfile

### PR DESCRIPTION
**Pull request by misha-kis**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM python:3.7-bullseye
```
Comment (kirill-ivanov-a) [inline Dockerfile]: Do we need the ability to run the algorithm on the GPU? If yes, then you should look at the [NVIDIA Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/overview.html)

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+COPY . .
+
+RUN pip install -U pip \
+    && pip install -r requirements.txt --no-cache-dir \
+    && cd dss_crf \
+    && python setup.py install \
+    && cd ..
+
+ENTRYPOINT ["python", "infer.py", "--path_to_pretrained_model=/200.pth", "--input_dir=/input", "--output_dir=/output"]
```
Comment (kirill-ivanov-a) [inline Dockerfile]: Let's create a separate file `run.py` that will contain the default running logic so we can run docker without arguments. Please, [see](https://github.com/prime-slam/line-detection-association-dockers/blob/main/detectors/F-Clip/run.py). It also seems to me that it is worth creating an adapter file that will convert the input and output of the algorithm into a format that is convenient for us, such as, for example, [here](https://github.com/prime-slam/line-detection-association-dockers/blob/main/detectors/F-Clip/adapter.py)

File: README.md [hunk 3]
```diff
@@ -21,35 +21,43 @@ If you use this code or our dataset (including test set), please cite:
...
+docker run -d --name gdnet_container gdnet
 ```
 
-### Usage
-First you need to download trained model at [here](https://mhaiyang.github.io/CVPR2020_GDNet/index.html).
+It is necessary to copy model and input to the container.
+You can download trained model [here](https://mhaiyang.github.io/CVPR2020_GDNet/index.html).
+To copy the pretrained model and input directory to the container, run
+```
+docker cp MODEL.PTH gdnet_container:/200.pth
```
Comment (kirill-ivanov-a) [inline README.md]: It is better to install all dependencies directly during the docker build. So it's worth adding layers with model loading to the Dockerfile. It is worth paying attention to [gdown](https://github.com/wkentaro/gdown). An example of downloading a model during build can be viewed [here](https://github.com/prime-slam/line-detection-association-dockers/blob/main/detectors/F-Clip/Dockerfile)

File: README.md [hunk 4]
```diff
@@ -21,35 +21,43 @@ If you use this code or our dataset (including test set), please cite:
...
+docker cp INPUT_DIR dgnet_container:/input
+```
 
-Then you can run `infer.py`:
+To run the container, run
+```
+docker start gdnet_container
```
Comment (kirill-ivanov-a) [inline README.md]: There is no need to stop your container, so it's better to start it in one-shot manner, that is, delete it after it has finished: `docker run --rm`

File: README.md [hunk 5]
```diff
@@ -21,35 +21,43 @@ If you use this code or our dataset (including test set), please cite:
...
 ```
 
-### Usage
-First you need to download trained model at [here](https://mhaiyang.github.io/CVPR2020_GDNet/index.html).
+It is necessary to copy model and input to the container.
+You can download trained model [here](https://mhaiyang.github.io/CVPR2020_GDNet/index.html).
+To copy the pretrained model and input directory to the container, run
+```
+docker cp MODEL.PTH gdnet_container:/200.pth
+docker cp INPUT_DIR dgnet_container:/input
```
Comment (kirill-ivanov-a) [inline README.md]: What happens if we have a very large input? It is better to [mount](https://docs.docker.com/storage/bind-mounts/) the directory directly when starting the container. An example can be found in [README](https://github.com/prime-slam/line-detection-association-dockers/tree/main/detectors/F-Clip)

File: README.md [hunk 6]
```diff
@@ -21,35 +21,43 @@ If you use this code or our dataset (including test set), please cite:
...
 
-Then you can run `infer.py`:
+To run the container, run
+```
+docker start gdnet_container
 ```
-python infer.py --path_to_pretrained_model=PATH_TO_PRETRAINED_MODEL --input_dir=PATH_TO_DIR_WITH_IMAGES
+To copy the results from the container, run:
 ```
+docker cp gdnet_container:/output OUTPUT_DIR
```
Comment (kirill-ivanov-a) [inline README.md]: See comment about copying input

**Review by kirill-ivanov-a — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 1 above)
Comment (misha-kis) [reply Dockerfile]: The original script used the GPU for evaluation, however @vnmsklnk modified it to use CPU if the GPU is unavailable

**Comment by misha-kis** [comment]
All comments are currently fixed except for GPU support

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,16 @@
+FROM python:3.7-bullseye
```
Comment (kirill-ivanov-a) [inline Dockerfile]: Seems like this line is not needed

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+FROM pytorch/pytorch:1.0.1-cuda10.0-cudnn7-runtime
+
+WORKDIR /
```
Comment (kirill-ivanov-a) [inline Dockerfile]: Let's make a more descriptive name, for example `/detector`

File: infer.py [hunk 9]
```diff
@@ -89,11 +72,7 @@ def main():
             #                                       img_name[:-4] + "_h.png"))
             # Image.fromarray(f2).save(os.path.join(ckpt_path, exp_name, '%s_%s' % (exp_name, args['snapshot']),
             #                                       img_name[:-4] + "_l.png"))
-            Image.fromarray(f3).save(os.path.join(opt.output_dir, img_name))
+            Image.fromarray(f3).save(os.path.join(output_dir, img_name))
 
         end = time.time()
         print("Average Time Is : {:.2f}".format((end - start) / len(img_list)))
```
Comment (kirill-ivanov-a) [inline infer.py]: It seems that we still need an adapter that will also include a unified way to measure time. Let's make an adapter that will include the same steps for all algorithms: reading data, creating a model, preprocessing the data, getting results from the model, postprocessing the data, and saving the results. In this case, we will be able to guarantee that the algorithms were run under the same conditions. This is how it should look like: [torch_adapter](https://github.com/prime-slam/line-detection-association-dockers/blob/549a849ea048d5023790e9286bb05afe2de653d1/detectors/common/adapter/torch_adapter.py), [adapter for algo](https://github.com/prime-slam/line-detection-association-dockers/blob/549a849ea048d5023790e9286bb05afe2de653d1/detectors/F-Clip/adapter.py)

**Review by kirill-ivanov-a — CHANGES_REQUESTED** [review]

File: common/metrics_evaluator.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
+#  @Date    : 2023-03-19
+#  @Editor  : Mikhail Kiselyov
+#  @E-mail  : kiselev.0353@gmail.com
+#  Provided as is
+
+import os
+from pathlib import Path
+import json
```
Comment (kirill-ivanov-a) [inline common/metrics_evaluator.py:6]: Let's sort import in the following order: external and system `import`, external and system `from`,  local `import`, local `from`. Each group should be separated by a new line

File: common/metrics_evaluator.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+class MetricsEvaluator:
+    def __init__(self,
+                 prediction_dir: Path,
+                 ground_truth_dir: Path,
+                 output_path: Path):
+        self.prediction_dir = prediction_dir
+        self.ground_truth_dir = ground_truth_dir
+        self.output_path = output_path
+
+        self.pred_files = os.listdir(prediction_dir)
```
Comment (kirill-ivanov-a) [inline common/metrics_evaluator.py]: `sorted(prediction_dir.iterdir())`

File: common/network_runner_base.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
+#  @Date    : 2023-03-19
+#  @Editor  : Mikhail Kiselyov
+#  @E-mail  : kiselev.0353@gmail.com
+#  Provided as is
+
+
+import logging
```
Comment (kirill-ivanov-a) [inline common/network_runner_base.py:6]: Please, sort imports

File: common/network_runner_base.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+                 output_dir: Path,
+                 log_path: Path,
+                 model_path: Path):
+        self.input_dir = input_dir
+        self.output_dir = output_dir
+        self.log_path = log_path
+
+        logging.basicConfig(filename=log_path, encoding='utf-8', level=logging.INFO)
+
+        if not os.path.exists(output_dir):
```
Comment (kirill-ivanov-a) [inline common/network_runner_base.py]: `output_dir.mkdir(parents=True, exist_ok=True)`

File: network_runner.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
+#  @Date    : 2023-03-19
+#  @Editor  : Mikhail Kiselyov
+#  @E-mail  : kiselev.0353@gmail.com
+#  Provided as is
+
+import logging
```
Comment (kirill-ivanov-a) [inline network_runner.py:6]: Sort imports

File: common/network_runner_base.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+            os.makedirs(output_dir)
+
+        self._load_model(model_path)
+
+    def run(self):
+        for img_name in tqdm(os.listdir(self.input_dir)):
```
Comment (kirill-ivanov-a) [inline common/network_runner_base.py]: `for img_name in tqdm(sorted(self.input_dir.iterdir())):`

File: network_runner.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+            img = img.convert('RGB')
+            logging.info(f'Image {img_name} is a gray image. Converting to RGB.')
+
+        return img, img.size
+
+    def _write_img(self, img_name, prediction):
+        f1, f2, f3 = prediction
+        logging.info(f'Image {img_name} processed. Writing results.')
+        if self.calculate_secondary:
+            Image.fromarray(f1).save(self.output_dir / Path(img_name[:-4] + "_h.png"))
```
Comment (kirill-ivanov-a) [inline network_runner.py]: Maybe `img_name.name` will be better than `img_name[:-4]`. And what does "_h.png"/"_l.png" mean?

**Review by kirill-ivanov-a — CHANGES_REQUESTED** [review]

File: network_runner.py (hunk 16 above)
Comment (misha-kis) [reply network_runner.py]: I do not know. Actually, saving of these images was commented out in the original code. However, I was not sure that I needed to remove them, so I saved them, though not fully understanding their meaning.

**Review by kirill-ivanov-a — APPROVED** [review]

File: common/metrics_evaluator.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
+#  @Date    : 2023-03-19
```
Comment (anastasiia-kornilova) [inline common/metrics_evaluator.py:1]: Let's put some license here, maybe Apache

File: common/network_runner_base.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
+#  @Date    : 2023-03-19
```
Comment (anastasiia-kornilova) [inline common/network_runner_base.py:1]: The same

**Pull request merged**